src/book/02-projects/01-nubots/03-build-system.mdx

```diff
@@ -5,3 +5,526 @@ title: Build System
+
+#### Roles
+
+Role files define which modules should be grouped together in order to form a single role (a binary that performs a specific role).
+
+Role files live in the `roles` folder and have a `.role` extension.
+
+Every role file must call the `nuclear_role()` function. Each argument to the `nuclear_role()` function is a C++ namespaced path to a module that should be included in the role. For example, if the `KinematicsConfiguration` module in `module/motion/KinematicsConfiguration` should be included in the role then `motion::KinematicsConfiguration` should appear as an argument to `nuclear_role()`.
+
+<Alert type="info">
+
+`module` is removed from the start of all module arguments as all modules must reside in the `modules` folder.
+
+</Alert>
+
+Ideally, all modules should be independent of each other and, as such, it should not matter which order modules are specified to `nuclear_role()`. In reality, there are a couple of modules which, if they are included in a role, should be passed to `nuclear_role()` first. These modules are:
+
+- `extension::FileWatcher`: Needed before all other modules so that the configuration and script systems can work properly
+- `support::SignalCatcher`: Catches exceptions to print out useful stack traces. Needs to come before all other modules to allow debugging during module installation
+- `support::logging::ConsoleLogHandler` or `support::logging::FileLogHandler`: Install before other modules so that log messages will get printed during module installation
+- `motion::KinematicsConfiguration`: Needs to come before other modules that require information about the robots kinematic configuration in order to set up properly
+
+Below is an example of a role file. Any line preceded by a `#` is a comment and will be ignored.
+
+```bash
+nuclear_role(
+  # FileWatcher, ConsoleLogHandler and Signal Catcher Must Go First. KinematicsConfiguration usually goes after these
+  # and without it many roles do not run
+  extension::FileWatcher
+  support::SignalCatcher
+  support::logging::ConsoleLogHandler
+  # This must come first as it emits config which many roles depend on (e.g. SensorFilter, WalkEngine)
+  motion::KinematicsConfiguration
+  support::configuration::GlobalConfig
+  support::configuration::NetworkConfiguration
+  platform::darwin::HardwareIO
+  platform::darwin::SensorFilter
+  support::NUsight
+)
+
+```
+
+#### Extensions
+
+[Extensions](https://nuclear.readthedocs.io/en/latest/extension.html) add new keywords to NUClear's dictionary, allowing you to be more expressive in writing [reactions](https://nuclear.readthedocs.io/en/latest/components.html#reactions).
```

Comment:
That extensions link just has a bunch of TODOs :P @TrentHouliston

Reply:
I suppose somehow those should get done sometime. I’ll see if I can write up those sooner rather than later